File: andorid/app/src/main/java/com/gaudiyakirtan/myapplication/models/Verse.kt
package com.gaudiyakirtan.myapplication.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * The verse's master text rendered into one script, per docs/data/verse.md > DisplayScript.
 * `text` line count must match the owning [Verse.sourceTextMaster] line count (invariant enforced
 * by the pipeline, not re-validated at decode time).
 */
@Serializable
data class DisplayScript(
    @SerialName("script_code") val scriptCode: String,
    @SerialName("standard") val standard: String? = null,
    @SerialName("text") val text: List<String> = emptyList()
)

/**
 * The ordered per-word glossary for one language rendered in one script, per
 * docs/data/verse.md > WordToWord. `words` is a list of `[headword, gloss]` pairs; reading order
 * is meaningful and must never be re-sorted.
 */
@Serializable
data class WordToWord(
    @SerialName("language_code") val languageCode: String,
    @SerialName("script_code") val scriptCode: String,
    @SerialName("standard") val standard: String? = null,
    @SerialName("words") val words: List<List<String>> = emptyList()
)

/**
 * One ordered stanza of a [Song]. See docs/data/verse.md (spec v1).
 *
 * Only ~22% of verses in the shipped corpus carry [wordToWords] / [translations] -- both default
 * to an empty list so absence never crashes decode or requires null-checking at every call site.
 */
@Serializable
data class Verse(
    @SerialName("verse_number") val verseNumber: Int,
    @SerialName("source_text_master") val sourceTextMaster: List<String>,
    @SerialName("display_scripts") val displayScripts: List<DisplayScript> = emptyList(),
    @SerialName("word_to_words") val wordToWords: List<WordToWord> = emptyList(),
    @SerialName("translations") val translations: List<Translation> = emptyList()
)

File: web/src/models/Verse.ts
// Conforms to docs/data/verse.md, spec v1.
import { LanguageCode, ScriptCode, TransliterationStandard } from './Common'
import { ITranslation } from './Translation'

/** The verse's lines rendered in one script. Line count matches sourceTextMaster. */
export interface IDisplayScript {
  scriptCode: ScriptCode
  /** Only present when scriptCode === 'Latn'; absent/null otherwise. */
  standard?: TransliterationStandard | null
  text: string[]
}

/** The ordered per-word glossary for one language rendered in one script. */
export interface IWordToWord {
  languageCode: LanguageCode
  /** Script the headwords are written in. */
  scriptCode: ScriptCode
  standard?: TransliterationStandard | null
  /** Ordered [headword, gloss] pairs, in reading order. Never sorted. */
  words: [string, string][]
}

/**
 * One ordered stanza of a Song. Holds the master ISO 15919 Latin text, every script
 * rendering of it, the word-to-word glossary, and full Translations.
 */
export interface IVerse {
  /** 1-based printed label. Display only - never used to reorder; order = array order. */
  verseNumber: number
  /** Master lines in ISO 15919 Latin, with inline [FLAG_*] markers. Not for display. */
  sourceTextMaster: string[]
  /** The master rendered into each supported script, flags resolved. */
  displayScripts: IDisplayScript[]
  /** Per-word glossary. Absent on ~80% of verses in the shipped corpus - omit, don't stub. */
  wordToWords?: IWordToWord[]
  /** Full-meaning renderings. Absent on ~78% of verses in the shipped corpus. */
  translations?: ITranslation[]
}
